### ISSUE TITLE
PR#589 Causes chef-client installations to report as failed when they have actually succeeded

### ISSUE THREAD

**Issue by robcoward** [test-kitchen/test-kitchen#601]
#589 seems to be causing my converges to fail. I'm using bundler to pull in the latest github code for test-kitchen (windows-guest-support branch) & kitchen-ec2 (afiune/kitchen-ec2 Transport branch) and running 'bundle exec kitchen create' is launching an ec2 instance successfully, however when running the converge, it fails on trying to install chef-client. It makes no difference installing the latest omnibus or specifying an 11.x version.

Logging into the Win2012 server shows that Chef Client is in fact installed correctly.

```
-----> Installing Chef Omnibus (latest)
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #converge action: [Transport WinRM exited (1) using shell [powershell] for command: [
# @return True if chef path exist, False otherwise
function is_chef_installed { Test-Path /opscode/chef }

# @param $version the requested version of omnibus package
# @return True if omnibus needs to be installed, False otherwise
function should_update_chef($version) {
  if (-Not (is_chef_installed)) { return $true }

  try {
    $chef_version=(chef-solo -v).split(' ',2)[1]
  } catch [Exception] {
    $chef_version = ''
  }

  switch ($version) {
    { 'true', $chef_version -contains $_ } { return $false }
    'latest' { return $true }
    default { return $true }
  }
}

# @param $chef_url The url where we will download chef
# @param $chef_msi Temporal MSI path
function download_chef($chef_url, $chef_msi) {
  Write-Host -NoNewline "`r`tDownloading Chef..."
  (New-Object System.Net.WebClient).DownloadFile($chef_url, $chef_msi)
  Write-Host -NoNewline "`r`tDone!              "
}

# function to install chef with sort of a nice progress bar
function install_chef {
  $proc_msi = Start-Process -FilePath 'msiexec.exe' -ArgumentList "/qn /i $chef_msi" -Passthru
  $bar = ""
  while (-Not $proc_msi.HasExited ) {
    Write-Host -NoNewline "`r`t[MSI] [$bar"
    Start-Sleep 2
    $bar += "#"
  }

  if ($proc_msi.ExitCode -ne 0) {
    throw "msiexec was not successful. Received exit code $($proc_msi.ExitCode)"
  }
  rm -r $chef_msi
  Write-Host -NoNewline "`r`t[MSI] [$bar] Completed!
"
}

  $chef_msi = $env:systemdrive + "\chef.msi"

  If (should_update_chef latest) {
    Write-Host "-----> Installing Chef Omnibus (latest)
"
    download_chef "http://www.chef.io/chef/install.msi?" $chef_msi
    install_chef
  } else {
    Write-Host "-----> Chef Omnibus installation detected (latest)
"
  }
]
REMOTE ERROR:
msiexec was not successful. Received exit code 
At line:42 char:5
+     throw "msiexec was not successful. Received exit code 
$($proc_msi.ExitCode)"
```
It would appear as though Powershell is not actually returning anything in $proc_msi.ExitCode

I'm not a windows developer, but I thought that powershell could not be relied on to return an exit code under all circumstances ?

**Comment by jdmundrawala** [comment]
I'm not sure how you get a nil exit code

cc @smurawski any ideas?

**Comment by jdmundrawala** [comment]
Maybe we have to do a `WaitForExit()` before we can read the value

**Comment by robcoward** [comment]
A lot of examples of Start-Process that I can find online seem to use the -wait option. I'm guessing that wasn't done here in order to print a bar of #'s every 2 seconds ? I dont see that bar when running kitchen anyway, so is there any other benefit to not using -wait on the Start-Process that I'm not seeing here ?

**Comment by jdmundrawala** [comment]
I see the bar...maybe we should look into why that is not happening

**Comment by afiune** [comment]
@jdmundrawala @smurawski You don't have a `.ExitCode` if you don't use the `-Wait` flag.. That is why this is breaking..

**Comment by afiune** [comment]
Also I didn't use `-Wait` because that way I can't show the lovely Progress Bar!! :smile: 
````
[MSI] [#         ]
[MSI] [####      ]
[MSI] [######    ]
[MSI] [######### ]
````

**Comment by jdmundrawala** [comment]
@afiune I'm not sure that is true. It is working for me.

**Comment by afiune** [comment]
You might be able to use `$proc_msi.WaitForExit()` on line 39 (https://github.com/test-kitchen/test-kitchen/blob/4c0a713b3daf8143f5ebb0b262256488ca66d4f2/support/chef_helpers.ps1#L39) like you mentioned @jdmundrawala but someone need to test that..

**Comment by jdmundrawala** [comment]
@robcoward could you give that a shot

**Comment by jdmundrawala** [comment]
i may have been testing against the wrong thing... seeing problems now

**Comment by jdmundrawala** [comment]
progress bar no longer works correctly :(

**Comment by robcoward** [comment]
Out of interest, I've just tried https://github.com/nvmlabs/test-kitchen/commit/d0e80a925a9044a5eb24ce684f8c1f6ccaf9cc0d which gets me past the install chef client stage. I'll try  the WaitForExit() option in a bit @afiune. 

However I'm now being tripped up by the Powershell issue http://support.microsoft.com/kb/2842230 because the the AMI I'm using obviously doesnt have hotfix http://hotfixv4.microsoft.com/Windows%208%20RTM/nosp/Fix452763/9200/free/463941_intl_x64_zip.exe applied and is causing memory allocation errors when processing resources such as chef_gem[rest-client].

```
           ================================================================================
           Error executing action `install` on resource 'chef_gem[rest-client]'
           ================================================================================
           
           Mixlib::ShellOut::ShellCommandFailed
           ------------------------------------
           Expected process to exit with [0], but received '1'
           ---- Begin output of C:/opscode/chef/embedded/bin/gem install rest-client -q --no-rdoc --no-ri -v "1.7.2" ----
           STDOUT: 
           STDERR: C:/opscode/chef/embedded/lib/ruby/site_ruby/2.0.0/rubygems.rb:8:in `require': failed to allocate memory (NoMemoryError)
```
Should test-kitchen attempt to detect this issue and apply the patch, or at least warn the user ? I am happy to raise another issue for this if you think its worth it ?

**Comment by jdmundrawala** [comment]
lol...we should detect that in all our windows code. I'm :+1: on warning people about this sadness

**Comment by robcoward** [comment]
@afiune just tried the WaitForExit() call and it still fails with no exit code. Wonder if its in any way related to the Powershell bug ? There's no memory error reported that I can see but.......
```

**Comment by echohack** [comment]
Not sure if this helps, but I've only run into this issue on the latest versions of Vagrant (1.7.0+). Vagrant 1.6.5 seems to run `kitchen converge` on windows just fine with this branch. Validated this on both Windows 7 and Mac OS X 10 host machines with Vagrant 1.6.5, Virtualbox 4.3.22, and Test Kitchen 1.3.2.dev.

Note that if you upgrade to Vagrant 1.7.0+ or downgrade to Vagrant 1.6.5 that you will have to `kitchen destroy` your machines before they will converge successfully.

Vagrant changelog for reference: https://github.com/mitchellh/vagrant/blob/master/CHANGELOG.md

Let me know if you need me to verify anything.

**Comment by irabinovitch** [comment]
We're finding Vagrant 1.6.x works with test-kitchen and windows, but 1.7.0+ always fails.  Happy to help test anything here.

**Comment by robcoward** [comment]
@irabinovitch i dont believe that the version of vagrant is the issue here. Although i currently have v1.7.1 installed, at the time of reporting this issue i had v1.6.3, and the original report above specifies that this was observed with the kitchen-ec2 driver anyway, not vagrant.

**Comment by afiune** [comment]
@irabinovitch curious about that information. Did you tried the latest version from this branch in both scenarios? Any gist you can share about the error you are having? Thanks for your feedback. :smile:

**Comment by irabinovitch** [comment]
Apologies it looks like I commented on the wrong issue by mistake, must have gotten tabs confused.  We did run into this install issue as well though which is why I had the both up. Will up a new issue + gist for the vagrant 1.7 issues. Apologies for muddying the thread here.

**Comment by tyler-ball** [comment]
@fnichol just pushed to the windows-guest-support branch a large refactoring.  This should avoid this issue.  Will you re-open this issue if chef-client downloads are still reporting as successful even when they have failed?
